# BSF_genomics_pipeline
Analysis scripts, pipelines, and data processing code for the study â€œHuman activity shapes the current global distribution of wild and captive black soldier flies (Hermetia illucens)â€. This repository includes workflows for WGS data processing, variant filtering, genotype imputation, and population genomic analyses

## ğŸ“ Repository Structure

```
â”œâ”€â”€ 01_Variant calling - FastQC, Mapping, Variant calling, VCF filtering, and Phasing
â”œâ”€â”€ 02_Imputation scripts - GLIMPSE2, QUILT, and STITCH (Building a reference panel and imputation)
â””â”€â”€ 03_BSF_PopGen_Tutorial - A Jupyter notebook tutorial on the population genomics analysis
```

This repository contains all scripts and tools used in the population genomic analysis of wild and captive Hermetia illucens (Black Soldier Fly). The study combines newly generated whole-genome sequences from Kenya and Uganda with publicly available data to explore the speciesâ€™ global dispersal history, domestication signatures, and implications for breeding and sustainability.

Key analyses include:

    1.Variant calling and filtering

    2.Low-coverage imputation using QUILT

    3.Population structure (PCA, ADMIXTURE, Phylogenetic and TreeMix analyses)

    4.Genetic diversity (FST, Dxy, Tajimaâ€™s D, HET)

    5.Inbreeding and ROH

    6.LD decay and Ne

## ğŸ“¬ Contact
For questions or collaborations, please reach out to **[Peter Muchina]** at [kimanimuchina@gmail.com](mailto:your.email@domain.com)  